src/libraries/Common/tests/System/Net/Sockets/NameResolutionTestHelper.Unix.cs

```diff
@@ -0,0 +1,99 @@
+        /// In case of failure we may force the tests fail, so we can gather diagnostics from the CI environment.
+        /// After collecting enough information we shall alter call sites to use throwOnFailure:false,
+        /// since the failure does not indicate an error within the NCL product code.
+        /// </summary>
+        public static unsafe bool EnsureNameToAddressWorks(string hostName, ITestOutputHelper? testOutput, bool throwOnFailure)
+        {
+            addrinfo hint = default;
+            hint.ai_family = 0; // AF_UNSPEC
+            hint.ai_flags = 2; // AI_CANONNAME
+            hint.ai_socktype = 1; // SOCK_STREAM;
+
+            addrinfo* res = default;
+
+            int err = getaddrinfo(hostName, null, &hint, &res);
+            freeaddrinfo(res);
+
+            if (err != 0)
+            {
+                string failureInfo =
+                    $"Failed to resolve '{hostName}'! getaddrinfo error: {err}{Environment.NewLine}{LogUnixInfo()}";
+                testOutput?.WriteLine(failureInfo);
+                if (throwOnFailure)
+                {
+                    throw new Exception(failureInfo);
+                }
+
+                return false;
+            }
+
+            return true;
+        }
+
+        private static string LogUnixInfo()
+        {
+            StringBuilder bld = new StringBuilder();
+            bld.AppendLine($"Dns.GetHostName() == {Dns.GetHostName()}");
+            bld.AppendLine("--- /etc/hosts ---");
+
+            string etcHosts;
+            try
+            {
+                etcHosts = File.ReadAllText("/etc/hosts");
+            }
+            catch (Exception ex)
+            {
+                etcHosts = $"Failed to retrieve /etc/hosts: {ex.Message}";
+            }
+
+            string resolvConf;
+            try
+            {
+                resolvConf = File.ReadAllText("/etc/resolv.conf");
+            }
+            catch (Exception ex)
+            {
+                resolvConf = $"Failed to retrieve /etc/resolv.conf: {ex.Message}";
+            }
+
+            bld.AppendLine(etcHosts);
+            bld.AppendLine("--- /etc/resolv.conf ---");
+            bld.AppendLine(resolvConf);
+            bld.AppendLine("------");
+            return bld.ToString();
+        }
+
+#pragma warning disable 169
+        [StructLayout(LayoutKind.Sequential)]
+        unsafe struct addrinfo
+        {
+            public int ai_flags;
+            public int ai_family;
+            public int ai_socktype;
+            int ai_protocol;
+            uint ai_addrlen;
+            UIntPtr ai_addr;
+            UIntPtr ai_canonname;
+            addrinfo* ai_next;
+        };
+#pragma warning restore 169
+
+        [DllImport("libc")]
+        private static extern unsafe int getaddrinfo(string node, string service, addrinfo* hints, addrinfo** res);
+
+        [DllImport("libc")]
+        private static extern unsafe void freeaddrinfo(addrinfo* res);
```

Comment:
All of this P/Invoking is unlikely to be portable across various Unix implementations.  It's why we use C shims as an intermediary for the production source.

Reply:
I was suggesting to use `gethostbyname()` as it takes string returns NULL or result so it should be easy to check and be portable IMHO. We chat about it with @antonfirsov and the goal is to check if system can resolve it self without depending on production code we are trying to test. We do not need to process the answer in any detailed way. We could still move it to Interop to make it more portable.

Reply:
> I was suggesting to use `gethostbyname()`

@wfurt I've chosen `freeaddrinfo` because mostly because `gethostbyname` is marked obsolete. `freeaddrinfo` also provides a trivial way to also get the error code, which I find useful.

> All of this P/Invoking is unlikely to be portable across various Unix implementations. 

@stephentoub while this is true in general, I don't see any reason to worry about it in this particular case. The `addrinfo` structure seems ABI compatible [across](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/getaddrinfo.3.html) [unixes](http://man7.org/linux/man-pages/man3/getaddrinfo.3.html), and the lines responsible for the query [in our own PAL code](https://github.com/dotnet/runtime/blob/master/src/libraries/Native/Unix/System.Native/pal_networking.c#L279-L288) are also portable.

Reply:
> The addrinfo structure seems binary compatible across unixes

We've seen numerous cases where the actual layout differed in header files from what was shown in the man pages, including but not limited to when targeting 32-bit vs 64-bit.

> and the lines responsible for the query in our own PAL code are also portable.

Those are compiled by the C compiler, which sees the actual header files.  That's the primary reason the native shims exist.

Reply:
I also agree that we should not p/invoke directly but only use native C shims.

Reply:
Yes, I understand that @antonfirsov. But it gets you out of the structure business. Since it has pointers, the size will differ. socklen_t seems to be fixed to 32bit but I'm not sure if that is mandatory for al OSes. 
Shim as @stephentoub suggested would be portable path.

Reply:
OK, I understand the issue now.

From the various options I see, the simplest and safest one seems to be the use of `gethostbyname` as @wfurt suggested. Looks like there are several existing cases where we P/Invoke into similar simple methods in test code. Is there any strong counter-argument against using `gethostbyname`?

Reply:
If you mean the:
```
struct hostent *gethostbyname(const char *name);
```
function, and we treat the return as an opaque pointer, e.g.
```
public static extern IntPtr gethostbyname(string name);
```
I'm ok with that.